### PR TITLE
feat: RBF with decrease output

### PR DESCRIPTION
**Pull request by szymonlesisz**
depends on https://github.com/trezor/hd-wallet/pull/58

- `composeTransaction` additional params (skipPermutation, sequence)
- `composeTransaction` required utxo (used in hd-wallet/buildTx)
- `composeTransaction` skip utxo lower than dust limit (if not required)
- `composeTransaction` docs

- tests for `composeTransaction`
- tests for `replaceTransaction` (decrease output)

### CODE REVIEW

File: docs/methods/composeTransaction.md [hunk 1]
```diff
@@ -1,39 +1,65 @@
...
+* `account` — *required* `Object` containing essential data, partial result of [getAccountInfo method](getAccountInfo.md#result)
+    - `path` - *required* `string`
+    - `utxo` - *required* `Array`
+    - `addresses` - *required* `string`
+* `feeLevels` — *required* `Array` of objects. set of requested variants, partial result of `blockchainEstimateFee method`
+    - `feePerUnit` - *required* `string` satoshi per transaction byte.
+* `baseFee` — *optional* `number` base fee of transaction in satoshi. used in replacement transactions calculation (RBF) and DOGE
+* `floorBaseFee` — *optional* `boolean` decide whenever baseFee should be floored to the nearest baseFee unit, prevents from fee overpricing. used in DOGE
+* `sequence` — *optional* `number` transaction input field used in RBF or locktime transactions
+* `skipPermutation` — *optional* `boolean` do not sort calculated inputs/outputs (usage: RBF transactions)
```
Comment (goodhoko) [inline docs/methods/composeTransaction.md]: I would prefer to move these two sections to the respective sections above. Huh, that's not particularly clear. What I mean is that instead of
```
## Requests a payment to a set of given outputs.
## Precompose, prepare transaction to be signed by Trezor.
## Payment request params:
## Precompose params:
```

I'd much rather have
```
## Requests a payment to a set of given outputs.
## Payment request params:
## Precompose, prepare transaction to be signed by Trezor.
## Precompose params:
```

or better yet
```
## Requests a payment to a set of given outputs.
### Params
## Precompose, prepare transaction to be signed by Trezor.
### Params
```

File: docs/methods/composeTransaction.md [hunk 2]
```diff
@@ -1,39 +1,65 @@
...
+*Device and backend connection is not required for this case since all data are provided.*
+
+## Payment request params:
 [****Optional common params****](commonParams.md)
-* `outputs` — *obligatory* `Array` of recipients Objects described below
-* `coin` — *obligatory* `string` determines network definition specified in [coins.json](../../src/data/coins.json) file. Coin `shortcut`, `name` or `label` can be used.
-* `push` — *optional* `boolean` determines if composed transaction will be broadcasted into blockchain.
+* `outputs` — *required* `Array` of output objects described [below](#accepted-output-objects)
+* `coin` — *required* `string` determines network definition specified in [coins.json](../../src/data/coins.json) file. Coin `shortcut`, `name` or `label` can be used.
+* `push` — *optional* `boolean` determines if composed transaction will be broadcasted into blockchain network.
```
Comment (goodhoko) [inline docs/methods/composeTransaction.md]: Since this optional we should also mention what is the default value.

File: src/js/core/methods/ComposeTransaction.js [hunk 3]
```diff
@@ -59,12 +62,15 @@ export default class ComposeTransaction extends AbstractMethod {
         const { payload } = message;
         // validate incoming parameters
         validateParams(payload, [
-            { name: 'outputs', type: 'array', obligatory: true, allowEmpty: true },
```
Comment (goodhoko) [inline src/js/core/methods/ComposeTransaction.js:62]: Why do we remove this? Seems unrelated to the feature in question.

File: tests/__fixtures__/index.js [hunk 4]
```diff
@@ -65,7 +66,7 @@ let fixtures = [
     cardanoSignTransaction,
     // todo: missing fixtures: ChangePin.js
     // todo: missing fixtures: CipherKeyValue.js
-    // todo: missing fixtures: ComposeTransaction.js
+    composeTransaction,
```
Comment (goodhoko) [inline tests/__fixtures__/index.js:69]: 👏

File: tests/__txcache__/testnet/56ebd82cdc91548617a1756f8e271dedda94e7c4d267d3c6d4f65d2654e5f3e2.json [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    {
+      "prev_hash": "9347a0298bd43f69ba168674f28bd9b25698db76f4f925582d9816eac2a61937",
+      "prev_index": 0,
+      "script_sig": "160014026813d5323be9c5d5f2df64a2ed4f02f5873669",
+      "sequence": 4294967294
+    }
+  ],
+  "lock_time": 1906598,
+  "version": 2
+}
```
Comment (goodhoko) [inline tests/__txcache__/testnet/56ebd82cdc91548617a1756f8e271dedda94e7c4d267d3c6d4f65d2654e5f3e2.json]: Can we add a NL pls? (applies to `tests/__txcache__/testnet/5bf10db6244c703d7f831043e6b6b0d6f3974bb5cf9fd216397e7620dc2e1015.json` as well)

**Review by goodhoko — APPROVED** [review]
Also LTGM with some minor suggestions and questions.

**Review by tomasklim — APPROVED** [review]

File: src/js/core/methods/ComposeTransaction.js (hunk 3 above)
Comment (szymonlesisz) [reply src/js/core/methods/ComposeTransaction.js:62]: i've found it during writing test, composing transaction without defined outputs doesnt make sense and also will throw an error from hd-wallet.
validation error from connect will throw more meaningful message than hd-wallet

File: docs/methods/composeTransaction.md (hunk 1 above)
Comment (szymonlesisz) [reply docs/methods/composeTransaction.md]: 8c6abc07f4b9b245778c30b08c4d17d297687b50

File: docs/methods/composeTransaction.md (hunk 2 above)
Comment (szymonlesisz) [reply docs/methods/composeTransaction.md]: 8c6abc07f4b9b245778c30b08c4d17d297687b50

File: tests/__txcache__/testnet/56ebd82cdc91548617a1756f8e271dedda94e7c4d267d3c6d4f65d2654e5f3e2.json (hunk 5 above)
Comment (szymonlesisz) [reply tests/__txcache__/testnet/56ebd82cdc91548617a1756f8e271dedda94e7c4d267d3c6d4f65d2654e5f3e2.json]: be2a8c1b8909f35c8fcc690d6299da036d26a914

**Comment by szymonlesisz** [comment]
test are failing because of this:
0286d2f7769de24f44067ddbc735b9953ea93022

which is related to this:
https://github.com/trezor/hd-wallet/commit/f1f60fb7ee11f098739da5f79d3d88e99fe842db

once merged in hd-wallet i'll add another fixup commit here and rebase everything

File: docs/methods/composeTransaction.md (hunk 1 above)
Comment (goodhoko) [reply docs/methods/composeTransaction.md]: Thanks!

File: src/js/core/methods/ComposeTransaction.js (hunk 3 above)
Comment (goodhoko) [reply src/js/core/methods/ComposeTransaction.js:62]: allright. Thanks for clarification. .)
